[Gutenberg] Support POST requests

fluiddot opened the pull request:
**Related PRs:**
* https://github.com/WordPress/gutenberg/pull/49371
* https://github.com/wordpress-mobile/gutenberg-mobile/pull/5596
* Android
    * https://github.com/wordpress-mobile/WordPress-Android/pull/18181
    * https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2691
* iOS
    * https://github.com/wordpress-mobile/WordPressKit-iOS/pull/589
    * https://github.com/wordpress-mobile/WordPress-iOS/pull/20428

## Description

This PR adds support for POST requests coming from the editor.

## To test

**The editor doesn't make POST requests yet, hence this change can only be tested with a test PR: https://github.com/wordpress-mobile/WordPress-Android/pull/18202**

Alternatively, we can also follow testing instructions from https://github.com/Automattic/jetpack/pull/29756 to test a POST request in the VideoPress block.

## Regression Notes
1. Potential unintended areas of impact
This change shouldn't impact any unintended area. The only area that could be impacted is the network logic used in the editor.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
I tested that GET requests coming from the editor work as expected.

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [ ] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

fluiddot committed:
Add perform post request

fluiddot committed:
Update Gutenberg Mobile reference

fluiddot committed:
Fix checkstyle issue

fluiddot committed:
Bump FluxC library reference

fluiddot committed:
Update Gutenberg Mobile reference

fluiddot commented on build.gradle:
```diff
@@ -20,9 +20,9 @@ ext {
     automatticTracksVersion = '2.2.0'
-    gutenbergMobileVersion = 'v1.92.0-alpha1'
+    gutenbergMobileVersion = '5596-5a7826bee6a11d22a2a96703009afb1b2c47fba2'
     wordPressAztecVersion = 'v1.6.3'
-    wordPressFluxCVersion = 'trunk-fcede535112327abf59416b7b544cf526c44fd00'
+    wordPressFluxCVersion = '2691-a3bac6077768944a3e79a6a1eb4c7c22dd2e1183'
```
This references will be replaced once https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2691 and https://github.com/wordpress-mobile/gutenberg-mobile/pull/5596 are merged.

fluiddot committed:
Update Gutenberg Mobile reference

fluiddot commented on build.gradle:
```diff
@@ -20,9 +20,9 @@ ext {
+    gutenbergMobileVersion = '5596-966b12f8bd5fd7c35fb3ef499566451bfb6a2463'
     wordPressAztecVersion = 'v1.6.3'
-    wordPressFluxCVersion = 'trunk-fcede535112327abf59416b7b544cf526c44fd00'
+    wordPressFluxCVersion = '2691-a3bac6077768944a3e79a6a1eb4c7c22dd2e1183'
```
These references will be replaced once https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2691 and https://github.com/wordpress-mobile/gutenberg-mobile/pull/5596 are merged.

SiobhyB reviewed (approved):
@fluiddot, thank you for documenting the implementation so clearly! I was initially feeling intimidated about reviewing this set of PRs, but you outlined the changes in a way that was so easy to digest in https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2691 🙇‍♀️ 

I'm approving this PR based on the following:

- I went through the test plan outlined in https://github.com/wordpress-mobile/WordPress-Android/pull/18202 and verified I could go through all of the steps successfully using the installable build.
- I reviewed the code in this PR, https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2691, and the Android-specific code in https://github.com/WordPress/gutenberg/pull/49371. I verified the code follows the logic I'd expect, matching the same patterns as those previously defined for GET requests.

It may be a good idea for a native developer to review as a precaution, in case there's anything I missed.

fluiddot committed:
Merge branch 'trunk' into gutenberg/support-post-request

# Conflicts:
#	build.gradle

fluiddot commented:
> @fluiddot, thank you for documenting the implementation so clearly! I was initially feeling intimidated about reviewing this set of PRs, but you outlined in a changes in a way that was so easy to digest in [wordpress-mobile/WordPress-FluxC-Android#2691](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2691) 🙇‍♀️
> 
> I'm approving this PR based on the following:
> 
> * I went through the test plan outlined in #18202 and verified I could go through all of the steps successfully using the installable build.
> * I reviewed the code in this PR, [wordpress-mobile/WordPress-FluxC-Android#2691](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2691), and the Android-specific code in [WordPress/gutenberg#49371](https://github.com/WordPress/gutenberg/pull/49371). I verified the code follows the logic I'd expect, matching the same patterns as those previously defined for GET requests.

Thanks @SiobhyB for reviewing the PR 🙇 !

> It may be a good idea for a native developer to review as a precaution, in case there's anything I missed.

Yep, good idea. I'll see if an Android developer could take a quick look at the PR for a second review.

ovitrif reviewed (approved):
Looks like any other good Java code to my eyes 👍🏻

fluiddot committed:
Update Gutenberg Mobile reference

fluiddot committed:
Merge branch 'trunk' into gutenberg/support-post-request

# Conflicts:
#	build.gradle

fluiddot committed:
Point FluxC version to `trunk` commit

fluiddot committed:
Update Gutenberg Mobile reference with alpha tag

fluiddot committed:
Merge branch 'trunk' into gutenberg/support-post-request

# Conflicts:
#	build.gradle

fluiddot committed:
Use newer version of `wordPressFluxCVersion` library

fluiddot committed:
Use last `trunk` commit of `wordPressFluxCVersion` library

Updating `wordPressFluxCVersion` library to the last commit of `trunk` to include the changes related to POST requests and hotfix `2.23.1`.

fluiddot commented on build.gradle:
```diff
@@ -20,9 +20,9 @@ ext {
+    gutenbergMobileVersion = 'v1.93.0-alpha1'
     wordPressAztecVersion = 'v1.6.3'
-    wordPressFluxCVersion = '2.23.1'
+    wordPressFluxCVersion = 'trunk-8b2fc7fdc3fe4e864a065210b356c20f8882f043'
```
In order to include the change https://github.com/wordpress-mobile/WordPress-FluxC-Android/commit/4c38c0da2445f4136a92a587414e9e4a27e5e534 relate to this PR and the hotfix `2.23.1`, I updated the reference to https://github.com/wordpress-mobile/WordPress-FluxC-Android/commit/8b2fc7fdc3fe4e864a065210b356c20f8882f043 to include both.

cc @thomashorta as you created the hotfix in case you have any thoughts/concerns. Thanks!
